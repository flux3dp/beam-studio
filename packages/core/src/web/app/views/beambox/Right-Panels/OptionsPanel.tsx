import React, { useMemo } from 'react';

import ColorPanel from 'app/views/beambox/Right-Panels/ColorPanel';
import ImageOptions from 'app/views/beambox/Right-Panels/Options-Blocks/ImageOptions';
import InFillBlock from 'app/views/beambox/Right-Panels/Options-Blocks/InFillBlock';
import MultiColorOptions from 'app/views/beambox/Right-Panels/Options-Blocks/MultiColorOptions';
import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import PolygonOptions from 'app/views/beambox/Right-Panels/Options-Blocks/PolygonOptions';
import RectOptions from 'app/views/beambox/Right-Panels/Options-Blocks/RectOptions';
import TextOptions from 'app/views/beambox/Right-Panels/Options-Blocks/TextOptions';
import { getObjectLayer } from 'helpers/layer/layer-helper';
import { useIsMobile } from 'helpers/system-helper';

import styles from './OptionsPanel.module.scss';

interface Props {
  elem: Element;
  rx: number;
  polygonSides?: number;
  updateObjectPanel: () => void;
  updateDimensionValues: (val: { [key: string]: any }) => void;
}

function OptionsPanel({
  elem,
  rx,
  polygonSides,
  updateObjectPanel,
  updateDimensionValues,
}: Props): JSX.Element {
  const isMobile = useIsMobile();
  let contents: JSX.Element[];
  const showColorPanel = useMemo(() => {
    if (!elem) return false;
    if (!['rect', 'ellipse', 'path', 'text', 'polygon', 'g', 'use'].includes(elem?.tagName.toLowerCase())) return false;
    return getObjectLayer(elem as SVGElement)?.elem?.getAttribute('data-fullcolor') === '1';
  }, [elem]);
  if (elem) {
    const tagName = elem.tagName.toLowerCase();
    if (tagName === 'rect') {
      contents = [
        <RectOptions
          key="rect"
          elem={elem}
          rx={rx}
          updateDimensionValues={updateDimensionValues}
        />,
        showColorPanel ? (
          <ColorPanel key="color" elem={elem} />
        ) : (
          <InFillBlock key="fill" elem={elem} />
        ),
      ];
    } else if (tagName === 'polygon') {
      contents = [
        <PolygonOptions key="polygon" elem={elem} polygonSides={polygonSides} />,
        showColorPanel ? (
          <ColorPanel key="color" elem={elem} />
        ) : (
          <InFillBlock key="fill" elem={elem} />
        ),
      ];
    } else if (tagName === 'text') {
      contents = [
        <TextOptions
          key="text"
          elem={elem}
          textElement={elem as SVGTextElement}
          updateObjectPanel={updateObjectPanel}
          updateDimensionValues={updateDimensionValues}
          showColorPanel={showColorPanel}
        />,
        // eslint-disable-next-line no-nested-ternary
        showColorPanel ? (
          <ColorPanel key="color" elem={elem} />
        ) : isMobile ? (
          <InFillBlock key="fill" elem={elem} />
        ) : null,
      ];
    } else if (tagName === 'image' || tagName === 'img') {
      if (elem.getAttribute('data-fullcolor') === '1') contents = [];
      else
        contents = [<ImageOptions key="image" elem={elem} updateObjectPanel={updateObjectPanel} />];
    } else if (tagName === 'g' && elem.getAttribute('data-textpath-g')) {
      const textElem = elem.querySelector('text');
      contents = [
        <TextOptions
          key="textpath"
          isTextPath
          elem={elem}
          textElement={textElem}
          updateObjectPanel={updateObjectPanel}
          updateDimensionValues={updateDimensionValues}
        />,
      ];
    } else if (tagName === 'g') {
      contents = [
        showColorPanel ? (
          <MultiColorOptions key="multi-color" elem={elem} />
        ) : (
          <InFillBlock key="infill" elem={elem} />
        ),
      ];
    } else if (tagName === 'use') {
      contents = [showColorPanel ? <MultiColorOptions key="multi-color" elem={elem} /> : null];
    } else {
      contents = [
        showColorPanel ? (
          <ColorPanel key="color" elem={elem} />
        ) : (
          <InFillBlock key="infill" elem={elem} />
        ),
      ];
    }
  }

  return isMobile ? (
    <div className={styles.container}>
      <ObjectPanelItem.Divider />
      {contents?.reverse()}
    </div>
  ) : (
    <>
      {(contents) && (
        <div className={styles.panel}>
          <div className={styles.title}>OPTIONS</div>
          {contents}
      </div>
      )}
    </>
  );
}

export default OptionsPanel;
