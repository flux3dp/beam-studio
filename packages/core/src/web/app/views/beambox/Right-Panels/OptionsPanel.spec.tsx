import React from 'react';

import { render } from '@testing-library/react';

import OptionsPanel from './OptionsPanel';

const useIsMobile = jest.fn();

jest.mock('@core/helpers/system-helper', () => ({
  useIsMobile: () => useIsMobile(),
}));

jest.mock('@core/app/views/beambox/Right-Panels/Options-Blocks/ImageOptions', () => 'dummy-image-options');
jest.mock('@core/app/views/beambox/Right-Panels/Options-Blocks/InFillBlock', () => 'dummy-infill-block');
jest.mock('@core/app/views/beambox/Right-Panels/Options-Blocks/RectOptions', () => 'dummy-rect-options');
jest.mock('@core/app/views/beambox/Right-Panels/Options-Blocks/TextOptions', () => 'dummy-text-options');
jest.mock('@core/app/views/beambox/Right-Panels/Options-Blocks/PolygonOptions', () => 'dummy-polygon-options');
jest.mock('@core/app/views/beambox/Right-Panels/Options-Blocks/MultiColorOptions', () => 'dummy-multi-color-options');
jest.mock('@core/app/views/beambox/Right-Panels/ColorPanel', () => 'dummy-color-panel');
jest.mock('@core/app/views/beambox/Right-Panels/ObjectPanelItem');

const mockGetAttribute = jest.fn();

jest.mock('@core/helpers/layer/layer-helper', () => ({
  getObjectLayer: () => ({
    elem: {
      getAttribute: () => mockGetAttribute(),
    },
  }),
}));

describe('should render correctly', () => {
  beforeEach(() => {
    mockGetAttribute.mockReturnValue(null);
  });

  test('rect', () => {
    document.body.innerHTML = '<rect id="rect" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('rect')} />);

    expect(container).toMatchSnapshot();
  });

  test('text', () => {
    document.body.innerHTML = '<text id="text" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('text')} />);

    expect(container).toMatchSnapshot();
  });

  test('image', () => {
    document.body.innerHTML = '<image id="image" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('image')} />);

    expect(container).toMatchSnapshot();
  });

  describe('polygon', () => {
    test('desktop version', () => {
      document.body.innerHTML = '<polygon id="polygon" />';

      const { container } = render(<OptionsPanel elem={document.getElementById('polygon')} />);

      expect(container).toMatchSnapshot();
    });

    test('web version', () => {
      window.FLUX.version = 'web';
      document.body.innerHTML = '<polygon id="polygon" />';

      const { container } = render(<OptionsPanel elem={document.getElementById('polygon')} />);

      expect(container).toMatchSnapshot();
    });
  });

  test('others', () => {
    document.body.innerHTML = '<xxx id="xxx" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('xxx')} />);

    expect(container).toMatchSnapshot();
  });

  test('no element', () => {
    const { container } = render(<OptionsPanel elem={null} />);

    expect(container).toMatchSnapshot();
  });

  test('rect in full color layer', () => {
    mockGetAttribute.mockReturnValue('1');
    document.body.innerHTML = '<rect id="rect" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('rect')} />);

    expect(container).toMatchSnapshot();
  });
});

describe('should render correctly in mobile', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    useIsMobile.mockReturnValue(true);
  });

  test('rect', () => {
    document.body.innerHTML = '<rect id="rect" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('rect')} />);

    expect(container).toMatchSnapshot();
  });

  test('text', () => {
    document.body.innerHTML = '<text id="text" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('text')} />);

    expect(container).toMatchSnapshot();
  });

  test('image', () => {
    document.body.innerHTML = '<image id="image" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('image')} />);

    expect(container).toMatchSnapshot();
  });

  describe('polygon', () => {
    test('desktop version', () => {
      document.body.innerHTML = '<polygon id="polygon" />';

      const { container } = render(<OptionsPanel elem={document.getElementById('polygon')} />);

      expect(container).toMatchSnapshot();
    });

    test('web version', () => {
      window.FLUX.version = 'web';
      document.body.innerHTML = '<polygon id="polygon" />';

      const { container } = render(<OptionsPanel elem={document.getElementById('polygon')} />);

      expect(container).toMatchSnapshot();
    });
  });

  test('others', () => {
    document.body.innerHTML = '<xxx id="xxx" />';

    const { container } = render(<OptionsPanel elem={document.getElementById('xxx')} />);

    expect(container).toMatchSnapshot();
  });

  test('no element', () => {
    const { container } = render(<OptionsPanel elem={null} />);

    expect(container).toMatchSnapshot();
  });
});
