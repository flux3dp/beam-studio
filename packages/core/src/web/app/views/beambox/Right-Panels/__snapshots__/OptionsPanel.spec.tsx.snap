// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly image 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy ImageOptions
      <button
        type="button"
      >
        updateObjectPanel
      </button>
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile image 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
    <div>
      This is dummy ImageOptions
      <button
        type="button"
      >
        updateObjectPanel
      </button>
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile no element 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
  </div>
</div>
`;

exports[`should render correctly in mobile others 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
    <div>
      This is dummy InFillBlock
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile polygon desktop version 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
    <div>
      This is dummy InFillBlock
    </div>
    <div>
      This is dummy PolygonOptions
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile polygon web version 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
    <div>
      This is dummy InFillBlock
    </div>
    <div>
      This is dummy PolygonOptions
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile rect 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
    <div>
      This is dummy InFillBlock
    </div>
    <div>
      This is dummy RectOptions
      <button
        type="button"
      >
        updateDimensionValues
      </button>
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile text 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="adm-divider adm-divider-vertical adm-divider-center divider"
    />
    <div>
      This is dummy InFillBlock
    </div>
    <div>
      This is dummy TextOptions
      <button
        type="button"
      >
        updateDimensionValues
      </button>
      <button
        type="button"
      >
        updateObjectPanel
      </button>
    </div>
  </div>
</div>
`;

exports[`should render correctly no element 1`] = `<div />`;

exports[`should render correctly others 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy InFillBlock
    </div>
  </div>
</div>
`;

exports[`should render correctly polygon desktop version 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy PolygonOptions
    </div>
    <div>
      This is dummy InFillBlock
    </div>
  </div>
</div>
`;

exports[`should render correctly polygon web version 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy PolygonOptions
    </div>
    <div>
      This is dummy InFillBlock
    </div>
  </div>
</div>
`;

exports[`should render correctly rect 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy RectOptions
      <button
        type="button"
      >
        updateDimensionValues
      </button>
    </div>
    <div>
      This is dummy InFillBlock
    </div>
  </div>
</div>
`;

exports[`should render correctly rect in full color layer 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy RectOptions
      <button
        type="button"
      >
        updateDimensionValues
      </button>
    </div>
    <div>
      This is dummy ColorPanel
    </div>
  </div>
</div>
`;

exports[`should render correctly text 1`] = `
<div>
  <div
    class="panel"
  >
    <div
      class="title"
    >
      OPTIONS
    </div>
    <div>
      This is dummy TextOptions
      <button
        type="button"
      >
        updateDimensionValues
      </button>
      <button
        type="button"
      >
        updateObjectPanel
      </button>
    </div>
  </div>
</div>
`;
